## Read and understand the Data 

In [8]:
data = "/home/slark/DE-projects/Commercial-data/super_store/data/01_raw/store_data.csv"

In [5]:
%%bash 
pip install pandas


[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


In [6]:
import pandas as pd

In [9]:
df = pd.read_csv(data)

In [10]:
df.head()

,order_id,order_date,ship_date,ship_mode,customer_name,segment,state,country,market,region,...,category,sub_category,product_name,sales,quantity,discount,profit,shipping_cost,order_priority,year
0,AG-2011-2040,01/01/2011,06/01/2011,Standard Class,Toby Braunhardt,Consumer,Constantine,Algeria,Africa,Africa,...,Office Supplies,Storage,"Tenex Lockers, Blue",408,2,0.0,106.140,35.46,Medium,2011
1,IN-2011-47883,01/01/2011,08/01/2011,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Office Supplies,Supplies,"Acme Trimmer, High Speed",120,3,0.1,36.036,9.72,Medium,2011
2,HU-2011-1220,01/01/2011,05/01/2011,Second Class,Annie Thurman,Consumer,Budapest,Hungary,EMEA,EMEA,...,Office Supplies,Storage,"Tenex Box, Single Width",66,4,0.0,29.640,8.17,High,2011
3,IT-2011-3647632,01/01/2011,05/01/2011,Second Class,Eugene Moren,Home Office,Stockholm,Sweden,EU,North,...,Office Supplies,Paper,"Enermax Note Cards, Premium",45,3,0.5,-26.055,4.82,High,2011
4,IN-2011-47883,01/01/2011,08/01/2011,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Furniture,Furnishings,"Eldon Light Bulb, Duo Pack",114,5,0.1,37.770,4.70,Medium,2011


In [11]:
df.shape

#51290 rows
#21 columns

(51290, 21)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        51290 non-null  object 
 1   order_date      51290 non-null  object 
 2   ship_date       51290 non-null  object 
 3   ship_mode       51290 non-null  object 
 4   customer_name   51290 non-null  object 
 5   segment         51290 non-null  object 
 6   state           51290 non-null  object 
 7   country         51290 non-null  object 
 8   market          51290 non-null  object 
 9   region          51290 non-null  object 
 10  product_id      51290 non-null  object 
 11  category        51290 non-null  object 
 12  sub_category    51290 non-null  object 
 13  product_name    51290 non-null  object 
 14  sales           51290 non-null  object 
 15  quantity        51290 non-null  int64  
 16  discount        51290 non-null  float64
 17  profit          51290 non-null 

There are no nulls 

Updates to data types required: 
- order_date str type --> convert to datetime 
- ship_date str type --> convert to datetime 
- sales str type --> convert to float

## Validate and Clean the data 

### Datatype conversion

In [13]:
#converting date str to datetime
df['order_date'] = pd.to_datetime(df['order_date'], dayfirst=True)
df['ship_date'] = pd.to_datetime(df['ship_date'], dayfirst=True)
df.head()

,order_id,order_date,ship_date,ship_mode,customer_name,segment,state,country,market,region,...,category,sub_category,product_name,sales,quantity,discount,profit,shipping_cost,order_priority,year
0,AG-2011-2040,2011-01-01,2011-01-06,Standard Class,Toby Braunhardt,Consumer,Constantine,Algeria,Africa,Africa,...,Office Supplies,Storage,"Tenex Lockers, Blue",408,2,0.0,106.140,35.46,Medium,2011
1,IN-2011-47883,2011-01-01,2011-01-08,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Office Supplies,Supplies,"Acme Trimmer, High Speed",120,3,0.1,36.036,9.72,Medium,2011
2,HU-2011-1220,2011-01-01,2011-01-05,Second Class,Annie Thurman,Consumer,Budapest,Hungary,EMEA,EMEA,...,Office Supplies,Storage,"Tenex Box, Single Width",66,4,0.0,29.640,8.17,High,2011
3,IT-2011-3647632,2011-01-01,2011-01-05,Second Class,Eugene Moren,Home Office,Stockholm,Sweden,EU,North,...,Office Supplies,Paper,"Enermax Note Cards, Premium",45,3,0.5,-26.055,4.82,High,2011
4,IN-2011-47883,2011-01-01,2011-01-08,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Furniture,Furnishings,"Eldon Light Bulb, Duo Pack",114,5,0.1,37.770,4.70,Medium,2011


In [14]:
df[['order_date', 'ship_date']].dtypes

order_date    datetime64[ns]
ship_date     datetime64[ns]
dtype: object

In [15]:
#removing comma's for numbers in thousands and converting sales column type from str to int 
df['sales'] = df['sales'].replace({',':''}, regex=True).astype(float)
df['sales'].dtype

dtype('float64')

### Checking duplicates

In [16]:
#check for duplicates 

#new df for just duplicates of order and product id combination 
duplicates=df[df.duplicated(subset=['order_id', 'product_id'], keep=False)]  #keep=false to see all duplicates, keep='first' by default and shows subsequent duplicates
duplicates[['order_id', 'product_id', 'product_name', 'quantity']]

,order_id,product_id,product_name,quantity
1377,IN-2011-77892,TEC-PH-10001751,"Samsung Smart Phone, Cordless",2
1381,IN-2011-77892,TEC-PH-10001751,"Samsung Smart Phone, Cordless",5
1781,US-2011-150119,FUR-CH-10002965,Global Leather Highback Executive Chair with P...,2
1785,US-2011-150119,FUR-CH-10002965,Global Leather Highback Executive Chair with P...,2
1970,ID-2011-54176,OFF-BI-10000168,"Cardinal Hole Reinforcements, Recycled",3
...,...,...,...,...
49443,CA-2014-118017,TEC-AC-10002006,Memorex Micro Travel Drive 16 GB,8
49483,CA-2014-118017,TEC-AC-10002006,Memorex Micro Travel Drive 16 GB,6
49731,IT-2014-4809306,FUR-CH-10004943,"Office Star Bag Chairs, Adjustable",5
49752,IT-2014-4809306,FUR-CH-10004943,"Office Star Bag Chairs, Adjustable",2


In [17]:
duplicates.shape[0]

#38 duplicated rows (with corresponding order and product ids) out of 51290 
#76 rows in total

76

In [18]:
duplicates.groupby(['order_id', 'product_id']).size()

#only ever one repeat of this order_id + product_id combination

order_id         product_id      
AG-2014-4840     TEC-MEM-10002524    2
CA-2012-103135   OFF-BI-10000069     2
CA-2013-129714   OFF-PA-10001970     2
CA-2013-137043   FUR-FU-10003664     2
CA-2013-140571   OFF-PA-10001954     2
CA-2014-118017   TEC-AC-10002006     2
CA-2014-152912   OFF-ST-10003208     2
ES-2012-2740636  OFF-AR-10004884     2
ES-2012-3945862  OFF-AR-10003066     2
ES-2012-4293573  OFF-ST-10000580     2
ES-2012-5238758  OFF-AR-10000399     2
ES-2013-2926991  OFF-BI-10000179     2
ES-2014-1314291  OFF-ST-10004367     2
ES-2014-1679056  OFF-BI-10002799     2
ES-2014-5856066  OFF-BI-10001119     2
ID-2011-43977    OFF-BI-10000340     2
ID-2011-54176    OFF-BI-10000168     2
IN-2011-45048    TEC-PH-10001354     2
IN-2011-71739    OFF-LA-10002806     2
IN-2011-77892    TEC-PH-10001751     2
IN-2012-38657    FUR-CH-10002510     2
IN-2013-52804    OFF-AP-10000729     2
IN-2013-66265    OFF-AR-10003461     2
IN-2014-34317    OFF-BI-10003309     2
IN-2014-45174    FUR-BO-100045

In [19]:
# df.groupby(['order_id', 'product_id']).agg(
#     rows=('quantity', 'count'),
#     total_quantity=('quantity', 'sum')
# ).query("rows > 1")

In [20]:
duplicates[df.duplicated(keep=False)]
#No true duplicates - another field is always different 

/tmp/ipykernel_16927/1682169426.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  duplicates[df.duplicated(keep=False)]


,order_id,order_date,ship_date,ship_mode,customer_name,segment,state,country,market,region,...,category,sub_category,product_name,sales,quantity,discount,profit,shipping_cost,order_priority,year


Duplicate Observations: 

- Out of the 51290 rows there are 38 that have duplicated order_id and product_id. 
- There is only ever 1 duplicate of this order_id and product_id combination, and other elements such as quantity vary. 
- This suggests that these duplicates are legitimate transactions rather than true duplicates so no duplicates are being removed.  


### Verifying locations

Verify market, region, state

In [21]:
df['market'].unique()

array(['Africa', 'APAC', 'EMEA', 'EU', 'Canada', 'LATAM', 'US'],
      dtype=object)

In [22]:
df['region'].unique()

#North, South, East, West --> little ambiguous

array(['Africa', 'Oceania', 'EMEA', 'North', 'Canada', 'Southeast Asia',
       'Central', 'South', 'Caribbean', 'North Asia', 'East', 'West',
       'Central Asia'], dtype=object)

In [23]:
df['state'].unique()


array(['Constantine', 'New South Wales', 'Budapest', ..., 'Karaman',
       'Sikasso', 'Atsimo-Andrefana'], shape=(1094,), dtype=object)

### Location combinations

In [24]:
#Filter to Africa market
df_market_africa = df[df['market'] == 'Africa']

#view location headings
df_market_africa.groupby(['market', 'region', 'state']).size()
#df_market_africa[['market', 'region', 'state']].drop_duplicates()

#too many to view, get complete list and export 

market  region  state                  
Africa  Africa  Abia                       25
                Abuja Capital Territory     2
                Adamaoua                    7
                Adamawa                     7
                Akwa Ibom                   8
                                           ..
                Yobe                        4
                Zambezia                   13
                Zamfara                     3
                Ziguinchor                  4
                Zinder                      4
Length: 228, dtype: int64

In [25]:
#export locations to CSV

df_locations = df[['market', 'region', 'state']]

#checks the location combination and how many rows have each combination
#df_locations.groupby(['market', 'region', 'state']).size().to_csv('01_data_checks/location_validation.csv', index=True)   # index=True keeps the grouped by columns in (as index)
df_locations.groupby(['market', 'region', 'state'])\
    .size()\
    .reset_index(name='row_count')\
    .to_csv('01_data_checks/01_location_combination_list.csv', index=False) 

#size() returns series (headings all part of index) give number of rows with that combination
#reset_index() makes df with separate columns, new column header for new column


In [26]:
#alternative: checks the location combination 

# df_locations.drop_duplicates()\
#     .sort_values(['market', 'region', 'state'])\
#     .to_csv('01_data_checks/01_location_combination_list_1.csv', index=False)

### Checking each states relatetes to one region only 

In [26]:
#check each state only relates to one region

states_df = df_locations.groupby('state')['region']\
    .nunique()\
    .reset_index(name='number_of_regions')\
    .sort_values('number_of_regions', ascending=False)

#view any states that map to more than 1 region
states_df[states_df['number_of_regions'] > 1]

#Initially converted to csv then added comments manually 

,state,number_of_regions
211,Central,4
933,Southern,2
1007,Ulaanbaatar,2
863,San Juan,2
424,Huila,2
1034,Vienna,2
216,Centre,2
1010,Upper Austria,2
940,Styria,2
875,Santiago,2


### Detailing all states that map to more than one region in one combined table 

In [27]:
#count of rows per location combination (similar to above)
state_region_market_counts = (
    df_locations
        .groupby(['state', 'region', 'market'])
        .size()
        .reset_index(name='No_of_rows')
)

#count how many regions per state to check each state only relates to one region (similar to above)
region_counts = (
    state_region_market_counts
        .groupby('state')['region']
        .nunique()
        .reset_index(name='number_of_regions')
)

# join the above together so it can be viewed together
final_df = (
    state_region_market_counts
        .merge(region_counts, on='state')
        .sort_values(['number_of_regions','state', 'region', 'market'], ascending=False)
)

#filter to just the states with more than one region 
final_df = final_df[final_df['number_of_regions'] > 1]

final_df

#export to csv
final_df.to_csv("01_data_checks/01_state_check_2.csv", index=False)  #already created a file and updated manually, created this file separately for reference

### Checking each region maps to just one market 

In [28]:
# check each regions only relates to one market
regions_df = df.groupby('region')['market'].nunique().reset_index(name='number_of_markets').sort_values('number_of_markets', ascending=False)

regions_df = regions_df[regions_df['number_of_markets'] > 1]

regions_df



,region,number_of_markets
3,Central,3
10,South,3
7,North,2


- All regions that map to multiple markets have generic names (Central, South, North) which is acceptle
- Noted that region should not be viewed in isolation but in relation to the market

In [29]:
#count of rows per location combination (same)
state_region_market_counts = (
    df_locations
        .groupby(['state', 'region', 'market'])
        .size()
        .reset_index(name='No_of_rows')
)

#count how many markets per region to check each region only relates to one market
market_counts = (
    state_region_market_counts
        .groupby('region')['market']
        .nunique()
        .reset_index(name='number_of_markets')
)

# join the above together so it can be viewed together (same)
final_df = (
    state_region_market_counts
        .merge(market_counts, on='region')
        .sort_values(['number_of_markets', 'region', 'market'], ascending=False)
)

#filter to just the regionss with more than one market 
final_df = final_df[final_df['number_of_markets'] > 1]

final_df

#export to csv
final_df.to_csv("01_data_checks/01_region_check.csv", index=False)

Description: 
- Markets is the highest geographical level, broken down to region, then state (both nested)
<br><br>

Observations:
- Not all states are unique either due to generic terminology (eg. Central, Southern) or states in different regions having the same name (eg. San Juan in Argentina or Puerto Rico). Therefore states cannot be considered in isolation, but in relation to the region/market.
- It appears that not all states are inputted in the same language (eg. Central, Centrale, Centre or Ouest, Sud). This has shown up where there are duplicates, so indication that locations being listed in the same language may not be reliable throughout the data. 
- Regions are not unique (Generic terms eg.North, South) so should not be considered in isolation but in relation to the market/state
- The markets overlap in terms of geographic location, this is particularly ambiguous EMEA, where EU and Africa are also listed as individual markets. Reason for these groupings is unknown, it is possilbe that there is a business reason behind this eg. some product ranges being for particular regions only (EU/Africa) whereas others have a broader target audience (EMEA)
<br><br>

Assumptions/questions: 
- Assumed that these groupings are defined and understood by the business so not being reclassified at this stage.
- Would be helpful to understand the background of these location groupings from the business particularly where Markets overlap. 
<br><br>

Suggestions: 
- Use a logical location grouping eg. country/continent  
- Ensure location input is in the same language or agreed terminology used for specific locations 
- Possible to clean location further based on agreed groupings (suggestions may become clearer on further analysis) eg. removal of EMEA 

## Initial stats

In [30]:
df['quantity'].describe()

count    51290.000000
mean         3.476545
std          2.278766
min          1.000000
25%          2.000000
50%          3.000000
75%          5.000000
max         14.000000
Name: quantity, dtype: float64

In [32]:
df['discount'].describe()

count    51290.000000
mean         0.142908
std          0.212280
min          0.000000
25%          0.000000
50%          0.000000
75%          0.200000
max          0.850000
Name: discount, dtype: float64

- 75% of transactions have a discount of 20% or lower
- 50% of transactions have no discounta at all 
- max discount is much higher at 85% - outlier 
- Discount strategy? - does not appear uniformed
- relatively small % of orders have a wide range/high value discount 

In [33]:
df['discount'].describe([0.5,0.6,0.7,0.8,0.9,0.95,0.96,0.97,0.98,0.99])

count    51290.000000
mean         0.142908
std          0.212280
min          0.000000
50%          0.000000
60%          0.100000
70%          0.200000
80%          0.350000
90%          0.500000
95%          0.600000
96%          0.700000
97%          0.700000
98%          0.700000
99%          0.700000
max          0.850000
Name: discount, dtype: float64

In [34]:
df['profit'].describe()

count    51290.000000
mean        28.641740
std        174.424113
min      -6599.978000
25%          0.000000
50%          9.240000
75%         36.810000
max       8399.976000
Name: profit, dtype: float64

- High standard deviation
- significant losses to look at further 
- 25% transactions making no (or negative) profit 
- 50% transactions making low profit / median (9.24) is low 
- mean (28) > median (9) -skewed
- highest profit is well above average 
- distribution suggests that small number of high profit transactions may drive the large proportion of profit  

In [35]:
df['shipping_cost'].describe()

count    51290.000000
mean        26.375915
std         57.296804
min          0.000000
25%          2.610000
50%          7.790000
75%         24.450000
max        933.570000
Name: shipping_cost, dtype: float64

- again mean (26) > median (7) - average is not typical cost 
- min of 0 unclear. Could be justified (eg. promotion/included in product price/digital/internal transfer/refunded) or data could be missing
- very high max shiping cost, well above majority of orders 
- skewed suggesting small no of high freight costs order - exceptions to investigate - destinations/priority type/product type(bulky)?

In [36]:
df['sales'].describe()

count    51290.000000
mean       246.498440
std        487.567175
min          0.000000
25%         31.000000
50%         85.000000
75%        251.000000
max      22638.000000
Name: sales, dtype: float64

- mean (246) > median (85) - average is not typical
- high standard deviation 
- min of 0 - promotions/refunds? 
- max order is significantly higher than 75% of all orders
- again data suggests a skew of a small number of high value orders

Overall Observations: 
- further analysis needed into the how this data is skewed towards a small proportion of orders having high discount, logistics, profit and sales values
- In terms of the business question of increasing profitability will need to look at: 
    - discount strategy
    - loss drivers / profit drivers 
    - missing shipping costs / drivers for high shipping costs 
    - low revenue / high revenue products 

## Cleaned dataset

In [ ]:
#save clean data to csv: 01_cleaned_store_data.csv

#df.to_csv("/home/slark/DE-projects/Commercial-data/super_store/data/02_clean/01_cleaned_store_data.csv", index=False)

#when you convert to csv you lose formatting (eg. datetime)

In [1]:
%%bash

pip install pyarrow

#required for parquet



[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import pyarrow as pa

In [ ]:
df.to_parquet("/home/slark/DE-projects/Commercial-data/super_store/data/02_clean/01_cleaned_store_data.parquet", engine="pyarrow", compression="gzip", index=False)

Cleaned dataset ready for feature engineering